src/datasets/features.py

```diff
@@ -32,19 +35,154 @@
+        else:
+            raise ValueError(f"Unexpected timestamp object {arrow_type}.")
+    elif pyarrow.types.is_binary(arrow_type):
+        return "binary"
+    elif pyarrow.types.is_large_binary(arrow_type):
+        return "large_binary"
+    elif pyarrow.types.is_string(arrow_type):
+        return "string"
+    elif pyarrow.types.is_large_string(arrow_type):
+        return "large_string"
+    else:
+        raise ValueError(f"Arrow type {arrow_type} does not have a datasets dtype equivalent.")
+
+
+def string_to_arrow(datasets_dtype: str) -> pa.DataType:
+    """
+    string_to_arrow takes a datasets string dtype and converts it to a pyarrow.DataType.
+
+    In effect, `dt == string_to_arrow(_arrow_to_datasets_dtype(dt))`
+
+    This is necessary because the datasets.Value() primitive type is constructed using a string dtype
+
+    Value(dtype=str)
+
+    But Features.type (via `get_nested_type()` expects to resolve Features into a pyarrow Schema,
+        which means that each Value() must be able to resolve into a corresponding pyarrow.DataType, which is the
+        purpose of this function.
+    """
+    timestamp_regex = re.compile(r"^timestamp\[(.*)\]$")
+    timestamp_matches = timestamp_regex.search(datasets_dtype)
+    if timestamp_matches:
+        timestamp_internals = timestamp_matches.group(1)
+        internals_regex = re.compile(r"^(s|ms|us|ns),\s*tz=([a-zA-Z0-9/_+:]*)$")
+        internals_matches = internals_regex.search(timestamp_internals)
+        if timestamp_internals in ["s", "ms", "us", "ns"]:
+            return pa.timestamp(timestamp_internals)
+        elif internals_matches:
+            return pa.timestamp(internals_matches.group(1), internals_matches.group(2))
+        else:
+            raise ValueError(
+                f"{datasets_dtype} is not a validly formatted string representation of a pyarrow timestamp."
+                f"Examples include timestamp[us] or timestamp[us, tz=America/New_York]"
+                f"See: https://arrow.apache.org/docs/python/generated/pyarrow.timestamp.html#pyarrow.timestamp"
+            )
+    elif datasets_dtype not in pa.__dict__:
+        if str(datasets_dtype + "_") not in pa.__dict__:
             raise ValueError(
-                f"Neither {type_str} nor {type_str + '_'} seems to be a pyarrow data type. "
+                f"Neither {datasets_dtype} nor {datasets_dtype + '_'} seems to be a pyarrow data type. "
                 f"Please make sure to use a correct data type, see: "
                 f"https://arrow.apache.org/docs/python/api/datatypes.html#factory-functions"
             )
-        arrow_data_type_str = str(type_str + "_")
+        arrow_data_factory_function_name = str(datasets_dtype + "_")
     else:
-        arrow_data_type_str = type_str
+        arrow_data_factory_function_name = datasets_dtype
+
+    return pa.__dict__[arrow_data_factory_function_name]()
+
+
+@dataclass
+class Value:
```

Comment:
There is no behavior change, I just added the docstring below and moved the class up the file to be near the two conversion functions above!